File: 0542-01-matrix/0542-01-matrix.py
import copy

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        queue = deque()
        row_length, column_length  = len(mat), len(mat[0])
        for row in range(row_length):
            for column in range(column_length):
                if mat[row][column] == 0:
                    queue.append((row, column))
                else:
                    mat[row][column] = row_length * column_length
        
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        while queue:
            row, column = queue.popleft()
            for direction in directions:
                next_row, next_column = row + direction[0], column + direction[1]
                if (0 <= next_row < row_length and 0 <= next_column < column_length) and mat[next_row][next_column] == row_length * column_length:
                    mat[next_row][next_column] = mat[row][column] + 1
                    queue.append((next_row, next_column))
        return mat

File: 0542-01-matrix/0542-01-matrix.js
/**
 * @param {number[][]} mat
 * @return {number[][]}
 */
var updateMatrix = function(mat) {
  const chachQueue = [];
  const DIRECTIONS = [[0, 1], [0, -1], [-1, 0], [1, 0]];
  const TOTAL_COUNT = mat.length * mat[0].length;

  for (let row = 0; row < mat.length; row++) {
    for (let column = 0; column < mat[0].length; column++) {
      if (mat[row][column] === 0) {
        chachQueue.push([row, column]);
      } else {
        mat[row][column] = TOTAL_COUNT
      }
    }
  }

  while (chachQueue.length !== 0) {
    [currentposX, currentposY] = chachQueue.shift();

    DIRECTIONS.forEach(([directionPosX, directionPosY]) => {
      const [nextPosX, nextPosY] = [currentposX +  directionPosX, currentposY + directionPosY];
      const nextPos = mat[nextPosX]?.[nextPosY];

      if (nextPos && nextPos === TOTAL_COUNT) {
        mat[nextPosX][nextPosY] = mat[currentposX][currentposY] + 1;
        chachQueue.push([nextPosX, nextPosY]);
      }
    })
  }

  return mat
};